src/parser/core/Loggable.ts

```diff
@@ -0,0 +1,42 @@
+import {seededColor} from 'utilities/color'
+
+// Helper types for debug
+type LogParameters = Parameters<typeof console.log>
+type DebugCallback = (opts: {
+	/** Log the provided arguments. */
+	log: (...data: LogParameters) => void,
+}) => void
+
+export abstract class Loggable {
```

Comment:
```suggestion
export abstract class Debuggable {
```

I mean `log` is the only debug function atm but we could expand it out a bit in future. Few modules use it for debug semantics already i.e. `this.debug(() => addShitToTimeline())` for invuln's debug output